### ISSUE TITLE
Add support for debugger run-time behavior options

### ISSUE THREAD

**Issue by svaarala** [svaarala/duktape#288]
See https://github.com/svaarala/duktape/pull/286#issuecomment-133300079

**Comment by fatcerberus** [comment]
Regarding the comment here:
https://github.com/svaarala/duktape/pull/286#issuecomment-133300079

I like the second option too. :)
"Individual setting commands like SetPauseOnThrow true: cleaner, generates a lot of commands, extends well"

A bitmask is awkward for this, as then you're pigeonholing yourself into strictly binary flags, where sometimes a different type may be desired.  For example, PauseOnThrow could be a 3-way setting:
* 0: Don't pause on throw
* 1: Pause on uncaught
* 2: Pause on all

It's more intuitive, then, to have a SetPauseOnThrow command which simply takes an integer with one of these values.  Plus individual options can be changed in isolation this way, whereas with a bitmask the client would always have to send the full set of flags each time.

**Comment by fatcerberus** [comment]
So far my ideas for options are:
* OnThrow behavior: pause all, pause uncaught, never
* Tail call disable (more accurate callstack)

**Comment by svaarala** [comment]
Just to defend the bitmask option a bit - while it's not convenient for the debug client, it is convenient for the Duktape internals: instead of mapping a bunch of values into an internal structure one by one, you just copy a `duk_uint32_t` from the protocol into the heap structure. This is obviously best for keeping the Duktape footprint low. (I know the difference is small, but if one ignores things like this for *every* decision everywhere, suddenly all those small differences start to add up.)

If the approach to options would be a `SetConfig` commands, it could also be a mix, e.g. begin with a bitmask with options that are clearly binary, and follow up with dvalues for options that are not.

**Comment by svaarala** [comment]
Anyway, because options are a new mechanism I want to ensure it heads off in a good direction, i.e. is nice to maintain and is compact in design. A useful step to figure out the alternatives would be to come up with a rough list of potential config options that might creep in over time. What affects design is e.g. are there structures values like maps or lists which are a bit inconvenient to represent as dvalues (and might favor using multiple commands). If not, we can make do with bitmasks and/or individual numbers, booleans, and strings, which works reasonably with a single SetConfig command.

This kind of process is never 100% but it's still worth trying to do a reasonable effort on it. For example, I did 3 prototypes of the debugger protocol, including a message-based rather than stream-based protocol, to figure out where the problems are in practice. This was quite tedious but quite enlightening.

**Comment by fatcerberus** [comment]
Agreed on the potential for footprint creep.  It helps, of course, that the debugger is still experimental and thus not yet part of the versioning guarantees (although maintaining compatibility is still desirable, of course :) ), which gives a bit more leeway for new features.

**Comment by svaarala** [comment]
I've been mostly using the policy that "experimental" means changes are made if something unexpected comes up but there's pretty good effort to avoid the situation. The opposite policy would be to stuff experimental features to master quickly and then tweak them freely. That might work better than long term branches but has its own downsides. I don't actually mind which policy is used, but I like consistency when possible :-)

So it'd be a viable option to merge stuff and change it pretty freely in master, and then ensure things are in a good maintainable condition when it's time to release. This might break the work flow of some users though; I know several projects that build from Duktape master. While they don't expect it to be 100% stable, they also don't expect there to be a 50-50 chance of something changing.

One way to improve on this would be two development branches: master would be as it is now, but there'd be a more volatile branch for working on a release. But... there are downsides with that too, like if something there would be good enough to go to master but other stuff wasn't, there'd then be a need for cherry-picking etc. So for now I'm hoping a single master and carefully reviewed merges would be enough (most of my own feature development branches live for weeks or months before a merge, and they're rebased and cleaned up a number of times).

Another approach would be to keep master as now, but add a tag for indicating where to build if you want the latest semi-stable version. That tag would then be moved from time to time between releases.

**Comment by fatcerberus** [comment]
I like the way you handle master right now, stuff isn't generally merged until testcases have been written, documentation added, and just in general a lot of polishing done.  As a result it's a lot cleaner than most Git masters tend to be, and overall a very stable experience.

I'm presently using Duktape from master myself, by the way, specifically the master snapshots, and I'm comfortable using it for non-beta releases.  So I definitely wouldn't want to give up that stability. :)

**Comment by fatcerberus** [comment]
>The only problem with doing an arbitrary Eval might be that some functions are optimized with the assumption that there won't be a direct eval call in that function

(from https://github.com/svaarala/duktape/issues/280#issuecomment-134271209)

So with this in mind, a potential future option could be one that disables these optimizations.

**Comment by svaarala** [comment]
That would be too late for already compiled functions however; and when you stop debugging the functions would still continue to perform badly.

The biggest thing is probably avoiding the creation of an Arguments object, which will be needed if Eval is possible. Maybe it would be OK when actively debugging, but perhaps not OK if you have debugger support enabled but are not actively debugging.

On the other hand, enabling debugger support currently means keeping _Varmap around even when a debug session is not active. For many functions Duktape normally drops the _Varmap because the function can be proven not to need it; with debugger support enabled it will keep the _Varmap, regardless of whether debugging is currently active or not. But this is much less harmful than Arguments.

**Comment by fatcerberus** [comment]
Yeah, I read `arguments.rst` a while back, it was a fascinating read, to say the least.  Non-strict `arguments` is a very strange animal indeed.

But I see what you mean here.  The more invasive debug options that are added, the fewer assumptions the engine can make because the debugger could be attached at any time.  It's quite a balancing act.

**Comment by fatcerberus** [comment]
Not directly related but in the spirit of runtime configuration: there could be an API you can call on a heap as a promise that the debugger would never be attached to it, so that the compiler is free to do optimizations.

This would in fact be useful for minisphere, as I have a command line option for it, `--debug`, to allow a client to connect.  So it's not too nice that I already have to take a performance hit just by having debug support enabled when most of the time the `--debug` switch won't be used. :(

**Comment by svaarala** [comment]
Realistically, if you want a release build, turn off debugging support for that build. You'll need two binaries of course. It's quite difficult to simultaneously achieve good performance and have a run-time option for debugging.

**Comment by fatcerberus** [comment]
True, but having seen the Duktape codebase, it looks like most of the checks needed for such an option are outside of the hotpath.  The bytecode executor itself doesn't appear to care whether debugging is enabled.

I suppose compiler performance might be compromised, though.

The main problem for me is that minisphere is a general-purpose game engine using JS for game logic, so the line between release and debug builds is a bit blurred compared to a purpose-built engine.  There is very good reason to have debugging enabled even in the release build. :)

**Comment by svaarala** [comment]
If you disable the "executor interrupt" you'll gain a bit of performance - but you can't enable debugging at the same time.  The interrupt handling is in the hot path.

**Comment by fatcerberus** [comment]
Don't misunderstand me, I actually agree that a true release build should have debugging disabled, and I will probably do the same for future releases of minisphere for the redist engine.  This was just an idea I had to allow debug-enabled builds to have the debugger turned off for a session, to improve performance in "developer" builds when debugging isn't requested.
